File: src/repositories/user-repository.ts
import Database from 'better-sqlite3';
import { StoredUser } from '../domain/models.js';

function mapUser(row: Record<string, unknown>): StoredUser {
  return {
    id: String(row.id),
    email: String(row.email),
    passwordHash: String(row.password_hash),
    displayName: String(row.display_name),
    createdAt: String(row.created_at),
  };
}

export interface UserRepository {
  create(user: StoredUser): void;
  findByEmail(email: string): StoredUser | null;
  findById(id: string): StoredUser | null;
}

export function createUserRepository(db: Database.Database): UserRepository {
  return {
    create(user) {
      db.prepare(
        `
          INSERT INTO users (id, email, password_hash, display_name, created_at)
          VALUES (@id, @email, @passwordHash, @displayName, @createdAt)
        `,
      ).run(user);
    },
    findByEmail(email) {
      const row = db.prepare('SELECT * FROM users WHERE email = ?').get(email) as Record<string, unknown> | undefined;
      return row ? mapUser(row) : null;
    },
    findById(id) {
      const row = db.prepare('SELECT * FROM users WHERE id = ?').get(id) as Record<string, unknown> | undefined;
      return row ? mapUser(row) : null;
    },
  };
}

File: src/repositories/user_repository.py
import sqlite3

from src.domain.models import StoredUser


def _map_user(row: sqlite3.Row) -> StoredUser:
    return StoredUser(
        id=str(row["id"]),
        email=str(row["email"]),
        password_hash=str(row["password_hash"]),
        display_name=str(row["display_name"]),
        created_at=str(row["created_at"]),
    )


class UserRepository:
    def __init__(self, connection: sqlite3.Connection) -> None:
        self._connection = connection

    def create(self, user: StoredUser) -> None:
        self._connection.execute(
            """
            INSERT INTO users (id, email, password_hash, display_name, created_at)
            VALUES (?, ?, ?, ?, ?)
            """,
            (user.id, user.email, user.password_hash, user.display_name, user.created_at),
        )
        self._connection.commit()

    def find_by_email(self, email: str) -> StoredUser | None:
        row = self._connection.execute("SELECT * FROM users WHERE email = ?", (email,)).fetchone()
        return _map_user(row) if row else None

    def find_by_id(self, user_id: str) -> StoredUser | None:
        row = self._connection.execute("SELECT * FROM users WHERE id = ?", (user_id,)).fetchone()
        return _map_user(row) if row else None
